### PR TITLE
[FLINK-31445][runtime]Split resource allocate/release related logic from FineGrainedSlotManager to TaskManagerTracker

### PR DESCRIPTION
**Pull request by huwh**
## What is the purpose of the change
Currently the FineGrainedSlotManager is response to slots allocations and resources request/release. This makes the logical of FineGrainedSlotManager complicated, So we will move task manager related work from FineGrainedSlotManager to TaskManagerTracker, which already tracks task managers but not including request/release.

## Brief change log

  - *migrate tests to Junit5*
  - *move resource allocation/ resource declaration to TaskManagerTracker*
  - *move (un-)register task manager to TaskManagerTracker*
  - *move task manager idle release check to TaskManagerTracker*

## Verifying this change
  - *Add unit tests in FineGrainedTaskManagerTrackerTest*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

File: flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedTaskManagerTrackerTest.java [hunk 1]
```diff
@@ -18,144 +18,191 @@
...
+        assertThat(taskManagerTracker.getRegisteredTaskManagers()).isEmpty();
     }
 
-    @Test(expected = NullPointerException.class)
-    public void testRemoveUnknownPendingTaskManager() {
-        final FineGrainedTaskManagerTracker taskManagerTracker =
-                new FineGrainedTaskManagerTracker();
+    @Test
+    void testUnregisterUnknownTaskManager() {
+        assertThrows(
```
Comment (reswqa) [inline flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedTaskManagerTrackerTest.java:172]: Please use Assertions.assertThatThrowBy in assertJ.

File: flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultSlotStatusSyncerTest.java [hunk 2]
```diff
@@ -281,23 +263,34 @@ public void testSlotStatusProcessing() {
...
+                .hasValueSatisfying(
+                        taskManagerInfo ->
+                                assertThat(taskManagerInfo.getAvailableResource())
+                                        .isEqualTo(ResourceProfile.fromResources(3, 12)));
+        assertThat(taskManagerTracker.getAllocatedOrPendingSlot(allocationId1)).isPresent();
+        assertThat(taskManagerTracker.getAllocatedOrPendingSlot(allocationId2)).isNotPresent();
+        assertThat(taskManagerTracker.getAllocatedOrPendingSlot(allocationId3)).isPresent();
+        assertThat(taskManagerTracker.getAllocatedOrPendingSlot(allocationId1))
+                .hasValueSatisfying(
+                        slot -> assertThat(slot.getState()).isEqualTo(SlotState.ALLOCATED));
```
Comment (reswqa) [inline flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultSlotStatusSyncerTest.java:280]: These have ensured that the option presents a value, so `assertThat(taskManagerTracker.getAllocatedOrPendingSlot(allocationId1)).isPresent()` can be safely removed.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManagerRuntimeServices.java [hunk 3]
```diff
@@ -83,7 +83,13 @@ private static SlotManager createSlotManager(
                     slotManagerConfiguration,
                     slotManagerMetricGroup,
                     new DefaultResourceTracker(),
-                    new FineGrainedTaskManagerTracker(),
+                    new FineGrainedTaskManagerTracker(
+                            slotManagerConfiguration.getMaxTotalCpu(),
+                            slotManagerConfiguration.getMaxTotalMem(),
+                            slotManagerConfiguration.isWaitResultConsumedBeforeRelease(),
+                            slotManagerConfiguration.getTaskManagerTimeout(),
+                            slotManagerConfiguration.getDeclareNeededResourceDelay(),
```
Comment (reswqa) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManagerRuntimeServices.java:91]: How about introduce `TaskManagerTrackerConfiguration` and derive it from `SlotManagerConfiguration`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedSlotManager.java [hunk 4]
```diff
@@ -212,21 +171,13 @@ public void start(
 
         resourceManagerId = Preconditions.checkNotNull(newResourceManagerId);
         mainThreadExecutor = Preconditions.checkNotNull(newMainThreadExecutor);
-        resourceAllocator = Preconditions.checkNotNull(newResourceAllocator);
         resourceEventListener = Preconditions.checkNotNull(newResourceEventListener);
         slotStatusSyncer.initialize(
                 taskManagerTracker, resourceTracker, resourceManagerId, mainThreadExecutor);
         blockedTaskManagerChecker = Preconditions.checkNotNull(newBlockedTaskManagerChecker);
+        taskManagerTracker.initialize(newResourceAllocator, mainThreadExecutor);
```
Comment (reswqa) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedSlotManager.java:178]:
```suggestion
        taskManagerTracker.initialize(Preconditions.checkNotNull(newResourceAllocator), mainThreadExecutor);       
```
We'd better do this check before pass parameter, otherwise we have to rely on all implementations.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/TaskManagerTracker.java [hunk 5]
```diff
@@ -21,63 +21,62 @@
...
 import org.apache.flink.runtime.instance.InstanceID;
-import org.apache.flink.runtime.resourcemanager.WorkerResourceSpec;
 import org.apache.flink.runtime.resourcemanager.registration.TaskExecutorConnection;
-import org.apache.flink.runtime.util.ResourceCounter;
 
-import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Executor;
 
 /** Tracks TaskManager's resource and slot status. */
```
Comment (reswqa) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/TaskManagerTracker.java:29]: We need to describe this class in more detail, similar to the `TaskExecutorManager`.

**Review by reswqa — COMMENTED** [review]
Thanks @huwh for creating this.

Before making a comprehensive review, the following changes need to be made as the current approach will cause too much pollution to the history of git commits:
- All commits related to junit5 migration should be squashed to only one commt, you can refer to 00c2ebd3.
- All hot-fix commits should be placed at the top of the entire PR. In this PR, I can find a hot-fix commit after a non-hotfix commit 389c1c0a.
- 8e8359fc 603c7016 61a6fc42 can be squashed to only one commit. Because they actually do the similar work(Move some functions from `SlotManager` to `TaskManagerTracker`). And your current split is not entirely correct, some changes belonging to other commits have been mistakenly placed. IMO, Merging them into one does not affect the review and can maintain integrity. 

BTW, I have also left some comments, please take a look.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManagerRuntimeServices.java (hunk 3 above)
Comment (huwh) [reply flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManagerRuntimeServices.java:91]: Good idea, but these configuration are used both in DeclarativeSlotManager and FineGrainedSlotManager. How about create a new issue to do this after we remove DeclarativeSlotManager?

**Comment by huwh** [comment]
Thanks @reswqa , squash some commits, and address the first comments

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManagerRuntimeServices.java (hunk 3 above)
Comment (reswqa) [reply flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManagerRuntimeServices.java:91]: It makes sense to refactor this after `DeclarativeSlotManager` removed.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/ResourceDeclaration.java [hunk 6]
```diff
@@ -51,7 +51,7 @@ public int getNumNeeded() {
         return numNeeded;
     }
 
-    public Collection<InstanceID> getUnwantedWorkers() {
+    public Set<InstanceID> getUnwantedWorkers() {
```
Comment (reswqa) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/ResourceDeclaration.java:54]: What is the purpose of this change? It is best to add some description in the commit message.

File: flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedTaskManagerTrackerTest.java [hunk 7]
```diff
@@ -319,14 +435,118 @@ public void testGetStatistics() {
...
+                .isEqualTo(ResourceProfile.fromResources(6, 700));
+        assertThat(taskManagerTracker.getRegisteredResource()).isEqualTo(totalResource);
+        assertThat(taskManagerTracker.getNumberRegisteredSlots()).isEqualTo(10);
+        assertThat(taskManagerTracker.getNumberFreeSlots()).isEqualTo(8);
+        assertThat(taskManagerTracker.getPendingResource())
+                .isEqualTo(ResourceProfile.fromResources(4, 200));
+    }
+
+    @Test
+    void testTimeoutForUnusedTaskManager() throws Exception {
```
Comment (reswqa) [inline flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedTaskManagerTrackerTest.java:460]: It seems that `testTimeoutForUnusedTaskManager` in `FineGrainedSlotManagerTest` is not needed.

**Review by reswqa — COMMENTED** [review]
Thanks @huwh for the update, I left some comments, please take a look again~

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/ResourceDeclaration.java (hunk 6 above)
Comment (huwh) [reply flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/ResourceDeclaration.java:54]: unmodifiableCollection does not implement equals method. This make unit test hard to compare whether ResourceDeclaration is the same value.

also updated in commit message
